File: sample/src/main/java/com/hannesdorfmann/mosby/sample/mvp/lce/CountriesPresenter.java
package com.hannesdorfmann.mosby.sample.mvp.lce;

import android.util.Log;
import com.hannesdorfmann.mosby.mvp.MvpBasePresenter;
import java.util.List;

/**
 * @author Hannes Dorfmann
 */
public class CountriesPresenter extends MvpBasePresenter<CountriesView> {

  private static final String TAG = "CountriesPresenter";

  private int failingCounter = 0;
  private CountriesLoader countriesLoader;

  public void loadCountries(final boolean pullToRefresh) {

    Log.d(TAG, "loadCountries(" + pullToRefresh + ")");

    Log.d(TAG, "showLoading(" + pullToRefresh + ")");

    getView().showLoading(pullToRefresh);

    if (countriesLoader != null && !countriesLoader.isCancelled()) {
      countriesLoader.cancel(true);
    }

    countriesLoader = new CountriesLoader(++failingCounter % 2 != 0,
        new CountriesLoader.CountriesLoaderListener() {

          @Override public void onSuccess(List<Country> countries) {

            if (isViewAttached()) {
              Log.d(TAG, "setData()");
              getView().setData(countries);

              Log.d(TAG, "showContent()");
              getView().showContent();
            }
          }

          @Override public void onError(Exception e) {

            if (isViewAttached()) {

              Log.d(TAG, "showError("+e.getClass().getSimpleName()+" , " + pullToRefresh + ")");
              getView().showError(e, pullToRefresh);
            }
          }
        });
    countriesLoader.execute();
  }

  @Override public void onDestroy(boolean retainInstance) {
    super.onDestroy(retainInstance);

    if (!retainInstance && countriesLoader != null) {
      countriesLoader.cancel(true);
      Log.d(TAG, "onDestroy() --> cancel Loader");
    }
  }
}

File: sample-flow/src/main/java/com/hannesdorfmann/mosby3/sample/flow/countries/CountriesPresenter.kt
package com.hannesdorfmann.mosby3.sample.flow.countries

import android.util.Log
import com.hannesdorfmann.mosby3.mvp.MvpBasePresenter
import com.hannesdorfmann.mosby3.sample.flow.model.Atlas
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 *
 *
 * @author Hannes Dorfmann
 */
class CountriesPresenter @Inject constructor(val atlas: Atlas) : MvpBasePresenter<CountriesView>() {

  var subscription: Subscription? = null

  fun loadCountries(pullToRefresh: Boolean) {

    Log.d("Flow", "loadCountries($pullToRefresh)")

    view?.showLoading(pullToRefresh)

    subscription = atlas.getCountries()
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(
            {
              Log.d("Flow", "view.setData()")
              view?.setData(it)
            },
            {
              Log.d("Flow", "view.showError()")
              view?.showError(it, pullToRefresh)
            },
            {
              Log.d("Flow", "view.showContent()")
              view?.showContent()
            }
        )

  }

  override fun attachView(view: CountriesView?) {
    super.attachView(view)
    Log.d("Flow", "attaching View to $this")
  }

  override fun detachView(retainInstance: Boolean) {
    Log.d("Flow", "detaching View from $this retained: $retainInstance")
    Log.d("Flow", "-------------------")
    super.detachView(retainInstance)
    if (!retainInstance) {
      subscription?.unsubscribe()
    }
  }
}